crates/env/src/engine/on_chain/impls.rs

```diff
@@ -546,7 +548,57 @@ impl TypedEnvBackend for EnvInstance {
 
     fn instantiate_contract<E, ContractRef, Args, Salt, RetType>(
         &mut self,
-        params: &CreateParams<E, ContractRef, Args, Salt, RetType>,
+        params: &CreateParams<E, ContractRef, LimitParamsV2<E>, Args, Salt, RetType>,
+    ) -> Result<
+        ink_primitives::ConstructorResult<
+            <RetType as ConstructorReturnType<ContractRef>>::Output,
+        >,
+    >
+    where
+        E: Environment,
+        ContractRef: FromAccountId<E>,
+        Args: scale::Encode,
+        Salt: AsRef<[u8]>,
+        RetType: ConstructorReturnType<ContractRef>,
+    {
+        let mut scoped = self.scoped_buffer();
+        let ref_time_limit = params.ref_time_limit();
+        let proof_time_limit = params.proof_time_limit();
+        let storage_deposit_limit = params
+            .storage_deposit_limit()
+            .map(|limit| &*scoped.take_encoded(limit));
+        let enc_code_hash = scoped.take_encoded(params.code_hash());
+        let enc_endowment = scoped.take_encoded(params.endowment());
+        let enc_input = scoped.take_encoded(params.exec_input());
+        // We support `AccountId` types with an encoding that requires up to
+        // 1024 bytes. Beyond that limit ink! contracts will trap for now.
+        // In the default configuration encoded `AccountId` require 32 bytes.
+        let out_address = &mut scoped.take(1024);
```

Comment:
We can't make use of `AccountId::max_encoded_len()` here instead of hard coding 1024?

Reply:
It's a good idea. We will need to:

1. Enable `max-encoded-len` feature
2. Add some bounds to the traits and impls
3. Derive `MaxEncodedLen` for `AccountId`

It's not that much to do, but I'm inclined to make that a separate PR to this, so we can properly test it and see e.g. contract size differences if any.

Reply:
https://github.com/paritytech/ink/pull/2128